### ISSUE TITLE
Investigate and fix any GDPR concerns where data is exposed over GraphQL

### ISSUE THREAD

**Issue by TCMeldrum** [UserOfficeProject/stfc-user-office-project#148]
split from #13

Investigate and fix any issues presented through the graphql endpoint (on our local set you can find the graphql gui on http://localhost:8081/graphql). For instance what data can be accessed through this endpoint? Is any of that data sensitive/personal?

/overwrite opened 2020-09-04

**Comment by TickleThePanda** [comment]
Waiting on #146

**Comment by martin-trajanovski** [comment]
@sc19ew just to mention something that might not be really connected with the issue but maybe we should disable the GraphQL playground in production because now it is available but there is no real need for it.

**Comment by sc19ew** [comment]
@martin-trajanovski I agree that is a good idea, I have opened a new issue for this [here](https://github.com/UserOfficeProject/stfc-user-office-project/issues/193)

**Comment by sc19ew** [comment]
Most of the queries use the authorised decorator to minimise the number of different roles that have permissions to access it. The decorator works by firstly checking to see if the user has an ApiAccessToken if they don’t then it checks to see if the user is logged in, then if there are no arguments for the decorator i.e. no specific role constraints have been defined, then the body of the query will run as normal, meaning that if there are no arguments then in theory a query will be accessible to a logged in USER.

**Comment by TeraMike** [comment]
This looks correct by my own admittedly limited understanding of the code.  

However a quick look at some of the queries shows that a lot of them, if they don't have specific role constraints, have some other checks in the query to check for access rights (e.g.: checking if user has readrights for a specific questionary. See `getQuestionary` or `getQuestionaryOrDefault`).  
Sometimes a role might not give access to a particular thing (e.g.: in `SEPQueries`, the `get` query has three roles specified, but if the user does not have an SEP id then the query returns null anyway), so this seems like a way of adding more granularity to the checks.

It might be worth checking that a given query, if it has no role constraints, has some other checks if the data involved doesn't want to be exposed but otherwise it seems reasonable enough, I think? Hopefully I'm not totally misunderstanding here.

**Comment by sc19ew** [comment]
List of queries that return personal data and are accessible when logged in as a USER.
(Query- Personal details returned )

**USER QUERIES** 
getUser - All user details from user table
getPreviousCollaborators -BasicUserDetails
getBasicUserDetailsByEmail -BasicUserDetails
getBasic - BasicUserDetails
byRef - BasicUserDetails
getAgent - BasicUserDetails

**VISITS QUERIES**
getRegistration - BasicUserDetails

**PROPOSAL QUERIES** 
byRef - Access to proposals by proposal id 
get - Access to proposals by proposal id 

**INSTRUMENT QUERIES** 
byRef - BasicUserDetails 
get - BasicUserDetails

**CALL QUERIES** 
byRef - BasicUserDetails 
get - BasicUserDetails
getAll - BasicUserDetails

BasicUserDetails is a fragment that contains:
   id
  firstname
  lastname
  organisation
  position
  created
  placeholder

**Comment by sc19ew** [comment]
The above queries are theoretically accessible to a user, however most of this data is not accessible in practise to a user.
The tests for these queries state that for:
Call Queries
- A user can get all calls.
The instrument scientist on specific calls basic user details are exposed when logged in as a USER when selecting a call, however as this is information that is available for users outside of this endpoint it is not a GDPR violation. 
In the example the specific instrument scientist is a test user and has empty fields that would otherwise be full.
![Screenshot (21)](https://user-images.githubusercontent.com/86953844/134017052-e635fe33-977e-4170-969b-5f14ad4d5dd1.png)

**Comment by mr-field** [comment]
To clarify, can users run queries like `getBasicUserDetailsByEmail` and get real data back? It might be a problem if any logged in user can see anyone's details with just their email address.

**Comment by TickleThePanda** [comment]
I don't think `getBasicUserDetailsByEmail` returning real details is an issue - it's the only way we can support adding people to a proposal and it's needed to show the details of the people you're adding.

**Comment by TickleThePanda** [comment]
The `getUser` is a concern for me as it can return all user details. Does `getBasic` do the same for all users, too?

**Comment by TickleThePanda** [comment]
All of the *CALL QUERIES* and other *USER QUERIES* than `getUser` and `getBasic` seem fine to me.

**Comment by sc19ew** [comment]
Shall I open a new issue to look at reducing the permissions on these queries to stop all users being able to access them?
